### DISCUSSION TITLE
Adding widgets in CtkTable

### DISCUSSION THREAD

**Discussion by Faiqyazdani** [Akascape/CTkTable#33] [General]
```
import customtkinter
from CTkTable import *

root = customtkinter.CTk()
values = [
    [1, 2, 3, 4, 5],
    [1, 2, 3, 4, 5],
    [1, 2, 3, 4, 5],
    [1, 2, 3, 4, 5],
    [1, 2, 3, 4, 5]
]

new_row_values = [11, 12, 13, 14, 15]
new_col_values = [11, 12, 13, 14, 15]

table = CTkTable(master=root, row=5, column=5)
table.pack(expand=True, fill="both", padx=20, pady=20)
table.add_row(index=2, values=[6, 7, 8, 9, 10])
table.add_column(index=3, values=[11, 12, 13, 14, 15])
table.edit_row(row=1, values=[11, 12, 13, 14, 15])
table.edit_column(column=2, values=[21, 22, 23, 24, 25])
button = customtkinter.CTkButton(table, text="Click Me")
table.add_widget(button, row=1, column=2)
root.mainloop()


```
How can i add widget like CtkCheckbox ,CtkEntry using edit_row and edit_column function in customtkinter i am getting error when i pass list in edit_row function ?

**Comment by Akascape** [comment]
@Faiqyazdani option menu is not implemented in ctktable. But you can get entry widget table by passing `write=True` in the table parameters. 
I recommend using [tksheet](https://github.com/ragardner/tksheet) for that purpose. CTkTable is specially designed for viewing/showing tabular data.

**Reply by Faiqyazdani** [reply]
i am not able to undertand how can add input field and ,labels and checkboxes in every row through tkshhet

**Reply by Faiqyazdani** [reply]
i tried to add write =True but got error _tkinter.TclError: wrong # args: should be ".!ctkframe3.!ctkcanvas2.!ctkframe.!ctktable.!ctkentry.!entry insert index text"

**Reply by Akascape** [reply]
@Faiqyazdani 
> i tried to add write =True but got error _tkinter.TclError: wrong # args: should be ".!ctkframe3.!ctkcanvas2.!ctkframe.!ctktable.!ctkentry.!entry insert index text"

Refer this example:
```python
import customtkinter
from CTkTable import *

root = customtkinter.CTk()

value = [["A","B","C","D","E"],[1,2,3,4,5],[0,0,0,0,0],[1,2,0,0,0],[1,2,3,4,5]]

table = CTkTable(master=root, row=5, column=5, values=value, write=True)
table.pack(expand=True, fill="both", padx=20, pady=20)
    
root.mainloop()
```

> i am not able to undertand how can add input field and ,labels and checkboxes in every row through tkshhet

Read the wiki of tksheet: https://github.com/ragardner/tksheet/wiki/Version-7

**Reply by Faiqyazdani** [reply]
is it possible to add checkbox using CtkTable ?

**Reply by Akascape** [reply]
@Faiqyazdani No, that is not implemented. CTktable is only for viewing tabular data and doing basic operations. Use tksheet

**Reply by Faiqyazdani** [reply]
i am actually not able to understand its documentation

**Reply by shazadhamad** [reply]
when using write=True, how can you update the values on the table and lock some of the values so they aren't changeable is this possible

**Reply by Akascape** [reply]
@shazadhamad Yes, try passing disabled parameter: `table.edit(row, column, disabled=True)`

**Reply by shazadhamad** [reply]
@Akascape ok but how can i get the values to save based on what the user inputs to that cell.

**Reply by shazadhamad** [reply]
and save them to a text file.

**Reply by Akascape** [reply]
@shazadhamad use the `.get()` method to get the values. Save the list to a text file.

**Reply by timdnewman** [reply]
Fyi for any brave googenaughts finding this like I did:  this changed at some point I think, so as at 22 may 2024 you are looking for state='disabled' instead of disabled=True

(But other than that it works like a charm)

**Comment by shazadhamad** [comment]
@Akascape yes but when write is enabled, hwo do iget the text that the user changes it to. I couldnt find this in the documentation

**Reply by Akascape** [reply]
@shazadhamad same method for write also. The value list changes automatically when the user writes in the entries, use the get method which returns all values. I hope there is no issue in this.

**Reply by shazadhamad** [reply]
@Akascape i set write to true and ran the code and typed into the boxes but it didnt change the values, should i have each thing set a a different variable. also do you have an example of it.

**Reply by Akascape** [reply]
@shazadhamad Don't worry, I will check this and provide you an example.

**Reply by shazadhamad** [reply]
thanks.

**Reply by shazadhamad** [reply]
also @Akascape  im unsure on how to save a list to a text file and upload it into the table.

**Reply by shazadhamad** [reply]
did u find it @Akascape.

**Comment by OLIVERFRANCISCO** [comment]
What you could do to add actions such as deleting, editing or viewing a record
What you could do is create a frame for each record
Create the root
and then create the container where the table will go
The header of the table can be done with CTkLabel, although it is cumbersome, it can work for what you want to do
and after doing a for to go through all the elements
in the for, make a frame called row so that when the other labels are created where the records and actions will go, each of these widgets must have a root of where it has to go

An example so you understand me
   
    `import customtkinter as ctk
    
    class TablaCortes:
        def __init__(self, master):
            self.master = master
            self.frame_inferior = ctk.CTkFrame(master)
            self.frame_inferior.pack(expand=True, fill="both", pady=(10, 10), padx=(10, 10))
            self.tabla_cortes()
    
        def tabla_cortes(self):
            self.values = []
            self.values.append(["Animal", "Nombre del corte", "Acciones"])
            ctk.CTkLabel(self.frame_inferior, text="Animal \t\t Corte \t\t Acciones", font=("Arial Bold", 14)).pack(anchor="w", pady=(10, 0))
            try:
                # Simulación de datos
                self.cortes = [
                    (1, "Cerdo", "Lomo", "Lomo de cerdo"),
                    (2, "Cerdo", "Costilla", "Costilla de cerdo"),
                    (3, "Cerdo", "Panceta", "Panceta de cerdo")
                ]
    
                for corte in self.cortes:
                    # Crear un frame para cada fila
                    row_frame = ctk.CTkFrame(self.frame_inferior)
                    row_frame.pack(fill="x", pady=(0, 5))
    
                    # Añadir los valores de las columnas
                    ctk.CTkLabel(row_frame, text=corte[1]).pack(side="left", padx=(10, 0))
                    ctk.CTkLabel(row_frame, text=corte[3]).pack(side="left", padx=(10, 0))
    
                    # Añadir el botón de acciones
                    ctk.CTkButton(row_frame, text="Prueba", command=lambda c=corte: print(f"Hola {c[3]}")).pack(side="right", padx=(10, 0))
    
            except Exception as e:
                print(e)
    
    if __name__ == "__main__":
        root = ctk.CTk()
        app = TablaCortes(root)
        root.mainloop()`
      
      
      ejemplo
      
![image](https://github.com/user-attachments/assets/ed746287-3ce1-4c6c-b2f6-8d0105b5d01c)
